File: kosql-execute-async-java/src/main/java/com/github/jacokoo/kosql/execute/async/java/KoSQL.java
package com.github.jacokoo.kosql.execute.async.java;

import com.github.jacokoo.kosql.build.BuildResult;
import com.github.jacokoo.kosql.build.Builder;
import com.github.jacokoo.kosql.build.Context;
import com.github.jacokoo.kosql.compose.Entity;
import com.github.jacokoo.kosql.compose.InnerTable;
import com.github.jacokoo.kosql.compose.Table;
import com.github.jacokoo.kosql.compose.result.Mapper;
import com.github.jacokoo.kosql.compose.statement.*;
import io.vertx.core.Future;
import io.vertx.sqlclient.*;
import kotlin.Pair;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public abstract class KoSQL {
    private static final Logger LOG = LoggerFactory.getLogger(KoSQL.class);

    private final ContextFactory factory;
    private final SqlClient client;
    private final Builder builder;

    public KoSQL(SqlClient client, Builder builder, ContextFactory factory) {
        this.client = client;
        this.builder = builder;
        this.factory = factory;
    }

    public <T> Future<T> run(WithInnerKoSQL<T> fn) {
        return openConnection(client).compose(conn ->
            fn.with(new InnerKoSQL(conn)).eventually(it -> conn.close())
        );
    }

    public <T> Future<T> tx(WithInnerKoSQL<T> fn) {
        return openConnection(client).compose(conn ->
            new InnerKoSQL(conn).tx(fn).eventually(it -> conn.close())
        );
    }

    private void closeConnection(SqlConnection conn) {
        conn.close().onFailure(it -> LOG.error("Failed to close connection", it));
    }

    protected abstract Future<SqlConnection> openConnection(SqlClient client);
    protected abstract <T, R extends Entity<T>> T getGeneratedKey(RowSet<Row> row, InnerTable<T, R> table);

    @FunctionalInterface
    public interface ContextFactory {
        Context create(Builder builder);
    }

    @FunctionalInterface
    public interface WithInnerKoSQL<T> {
        Future<T> with(InnerKoSQL ko);
    }

    public class InnerKoSQL implements Dao {
        private final SqlConnection connection;
        private boolean started = false;

        private InnerKoSQL(SqlConnection connection) {
            this.connection = connection;
        }

        private <T> Future<T> tx(WithInnerKoSQL<T> fn) {
            if (started) return fn.with(this);

            return connection.begin().compose(tx ->
                fn.with(this)
                    .onFailure(it -> tx.rollback())
                    .compose(it -> tx.commit().map(it))
            ).onComplete(it -> started = false);
        }

        @Override
        public <K, T extends Table<K, ? extends Entity<K>>> Future<Integer> execute(UpdateStatement<K, T> update) {
            var result = builder.build(update, factory.create(builder));
            return execute(result).map(SqlResult::rowCount);
        }

        @Override
        public Future<Integer> execute(DeleteStatement delete) {
            var result = builder.build(delete, factory.create(builder));
            return execute(result).map(SqlResult::rowCount);
        }

        @Override
        public <T> Future<Pair<T, Integer>> execute(InsertStatement<T> insert) {
            var result = builder.build(insert, factory.create(builder));
            return execute(result).map(rs -> {
                var pk = insert.getData().getTable().primaryKey();
                var key = pk.getType().getNullValue();
                if (pk.getAutoIncrement()) {
                    key = pk.getType().fromDb(getGeneratedKey(rs, insert.getData().getTable()));
                }
                return new Pair<>(key, rs.rowCount());
            });
        }

        @Override
        public <T> Future<Integer> execute(BatchInsertStatement<T> insert) {
            var result = builder.build(insert, factory.create(builder));
            return prepare(result).compose(st ->{
                var params = result.getParams().stream()
                    .map(it -> Tuple.from((List<?>)it))
                    .collect(Collectors.toList());

                return st.query().executeBatch(params)
                    .eventually(i -> st.close())
                    .map(SqlResult::rowCount);
            });
        }

        @Override
        public <T, C extends ColumnList> Future<List<T>> execute(SelectStatement<C> select, Mapper<T> mapper) {
            var result = builder.build(select, factory.create(builder));
            return prepare(result).compose(st ->
                st.query().execute(Tuple.tuple(result.getParams()))
                    .map(rows -> {
                        List<T> list = new ArrayList<>();
                        for (var row : rows) {
                            var cs = select.getData().getColumns();
                            var source = new com.github.jacokoo.kosql.compose.result.Row(row::getValue, select.getData().getColumns());
                            list.add(mapper.map(source));
                        }
                        return list;
                    })
                    .eventually(i -> st.close())
            );
        }

        private Future<PreparedStatement> prepare(BuildResult result) {
            if (LOG.isDebugEnabled()) {
                var args = result.getParams().stream().map(Object::toString).collect(Collectors.joining(", "));
                LOG.debug("execute: \nSQL: {} \nwith params: {}", result.getSql(), args);
            }
            return connection.prepare(result.getSql());
        }

        private Future<RowSet<Row>> execute(BuildResult result) {
            return prepare(result).compose(st ->
                st.query()
                    .execute(Tuple.tuple(result.getParams()))
                    .eventually(t -> st.close())
            );
        }
    }
}

File: kosql-vertx/src/main/kotlin/com/github/jacokoo/kosql/execute/vertx/KoSQL.kt
package com.github.jacokoo.kosql.execute.vertx

import com.github.jacokoo.kosql.build.*
import com.github.jacokoo.kosql.compose.Compose
import com.github.jacokoo.kosql.compose.Entity
import com.github.jacokoo.kosql.compose.Table
import com.github.jacokoo.kosql.execute.async.Use
import com.github.jacokoo.kosql.compose.result.Mapper
import com.github.jacokoo.kosql.compose.statement.*
import com.github.jacokoo.kosql.execute.async.Dao
import com.github.jacokoo.kosql.compose.result.Row as Row2
import io.vertx.kotlin.coroutines.awaitResult
import io.vertx.sqlclient.SqlClient
import io.vertx.sqlclient.*
import org.slf4j.LoggerFactory
import kotlin.coroutines.suspendCoroutine

class KoSQL<B: Builder> internal constructor(
    private val kf: KoSQLFactory<B>,
    private val conn: SqlConnection
): Compose(), Use, Dao {
    private var started: Boolean = false

    companion object {
        private val LOG = LoggerFactory.getLogger(KoSQL::class.java.name.substringBefore("\$Companion"))
    }

    suspend fun <T> tx(fn: suspend KoSQL<B>.() -> T): T {
        if (started) {
            return fn()
        }

        val tran = conn.begin()
        started = true
        try {
            val re = fn()
            tran.commit()
            started = false
            return re
        } catch (e: Exception) {
            tran.rollback()
            throw e
        }
    }

    override suspend fun <K, T : Table<K, Entity<K>>> execute(update: UpdateStatement<K, T>): Int
            = kf.builder.build(update, context()).let { execute(it).rowCount() }

    override suspend fun execute(delete: DeleteStatement): Int
            = kf.builder.build(delete, context()).let { execute(it).rowCount() }

    override suspend fun <T> execute(insert: InsertStatement<T>): Pair<T, Int>
            = kf.builder.build(insert, context()).let { result ->
        val rs = execute(result)

        val pk = insert.data.table.primaryKey()
        var key = pk.type.nullValue
        if (pk.autoIncrement) {
            key = pk.type.fromDb(kf.getGeneratedKey(rs))
        }
        key to rs.rowCount()
    }

    override suspend fun <T> execute(insert: BatchInsertStatement<T>): Int
            = kf.builder.build(insert, context()).let { result ->
        val st = prepare(result)
        val rs = awaitResult<RowSet<Row>> { st.query().executeBatch(result.params.map {p ->
            Tuple.tuple(p as List<*>)
        }, it)}
        rs.rowCount()
    }

    override suspend fun <T, R : ColumnList> execute(select: SelectStatement<R>, mapper: Mapper<T>): List<T> =
        kf.builder.build(select, context()).let { result ->
            val stream = prepare(result).createStream(20, result.tuple())
            suspendCoroutine { cont ->
                val list = mutableListOf<T>()
                stream.handler {
                    list.add(mapper.map(Row2({ i -> it.getValue(i) }, select.data.columns)))
                }.endHandler {
                    cont.resumeWith(Result.success(list))
                }
            }
        }

    private fun context(): Context = kf.contextFactory(kf.builder)
    private fun BuildResult.tuple(): Tuple = Tuple.tuple(params)
    private suspend fun prepare(result: BuildResult): PreparedStatement {
        if (LOG.isDebugEnabled) {
            LOG.debug("execute: {}", result.sql)
            LOG.debug("execute with params: {}", result.params.joinToString(prefix = "(", postfix = ")"))
        }
        return awaitResult { conn.prepare(result.sql, it) }
    }
    private suspend fun execute(result: BuildResult): RowSet<Row> {
        val st = prepare(result)
        return awaitResult { st.query().execute(result.tuple(), it) }
    }

}

abstract class KoSQLFactory<B: Builder>(
    private val client: SqlClient,
    internal val builder: B,
    internal val contextFactory: (B) -> Context = { DefaultContext(it) }
) {

    suspend fun <T> tx(fn: suspend KoSQL<B>.() -> T): T {
        val conn = openConnection(client)
        try {
            return KoSQL(this, conn).tx { fn() }
        } finally {
            conn.close()
        }
    }

    suspend fun <T> run(fn: suspend KoSQL<B>.() -> T): T {
        val conn = openConnection(client)
        try {
            return KoSQL(this, conn).fn()
        } finally {
            conn.close()
        }
    }

    abstract suspend fun openConnection(client: SqlClient): SqlConnection
    abstract fun getGeneratedKey(row: RowSet<Row>): Any
}